### PR TITLE
perf: Api optimizations

### PR DESCRIPTION
**Pull request by kripanshdevtron**
# Description

Fixes #

## How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

## Checklist:

* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] Does this PR requires documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have performed a self-review of my own code.
* [ ] I have commented my code, particularly in hard-to-understand areas.
* [x] I have tested it for all user roles.
* [ ] I have added all the required unit/api test cases.

## Does this PR introduce a user-facing change?

```release-note

```

### CODE REVIEW

File: api/module/ModuleRouter.go [hunk 1]
```diff
@@ -18,6 +18,7 @@ func NewModuleRouterImpl(moduleRestHandler ModuleRestHandler) *ModuleRouterImpl
 
 func (impl ModuleRouterImpl) Init(configRouter *mux.Router) {
 	configRouter.Path("").HandlerFunc(impl.moduleRestHandler.GetModuleInfo).Queries("name", "{name}").Methods("GET")
```
Comment (kripanshdevtron) [inline api/module/ModuleRouter.go:20]: why do we need this ??

File: api/restHandler/AppListingRestHandler.go [hunk 2]
```diff
@@ -437,6 +463,148 @@ func (handler AppListingRestHandlerImpl) FetchAppsByEnvironment(w http.ResponseW
...
+		return
+	}
+
+	validAppIds := make([]int, 0)
+	//for non super admin users
+	if !isActionUserSuperAdmin {
+		userEmailId := strings.ToLower(user.EmailId)
+		rbacObjectsForAllAppsMap := handler.enforcerUtil.GetRbacObjectsForAllApps()
+		rbacObjectToAppIdsMap := make(map[string]([]int))
+		rbacObjects := make([]string, len(rbacObjectsForAllAppsMap))
```
Comment (kripanshdevtron) [inline api/restHandler/AppListingRestHandler.go:502]: initiliaze with proper length value

File: api/chartRepo/ChartRepositoryRestHandler.go [hunk 3]
```diff
@@ -115,13 +116,35 @@ func (handler *ChartRepositoryRestHandlerImpl) GetChartRepoList(w http.ResponseW
 	handler.Logger.Infow("request payload, GetChartRepoList, app store")
 	res, err := handler.chartRepositoryService.GetChartRepoList()
 
+	if err != nil {
+		handler.Logger.Errorw("service err, GetChartRepoList, app store", "err", err, "userId", userId)
+		handler.Logger.Debug(res)
```
Comment (kripanshdevtron) [inline api/chartRepo/ChartRepositoryRestHandler.go:121]: improve logging

File: api/chartRepo/ChartRepositoryRestHandler.go [hunk 4]
```diff
@@ -115,13 +116,35 @@ func (handler *ChartRepositoryRestHandlerImpl) GetChartRepoList(w http.ResponseW
...
+		common.WriteJsonResp(w, err, nil, http.StatusInternalServerError)
+		return
+	}
+
 	err = handler.attributesService.UpdateKeyValueByOne(CHART_STORE_VISITED_COUNTER)
+	// ignoring error here since it shouldn't break the main call. logging it instead
+	handler.Logger.Errorw("service err, GetChartRepoList, app store, update visited counter", "err", err, "userId", userId)
```
Comment (kripanshdevtron) [inline api/chartRepo/ChartRepositoryRestHandler.go:128]: need to correct error msg

File: internal/sql/repository/AppListingRepository.go [hunk 5]
```diff
@@ -200,38 +236,153 @@ func (impl AppListingRepositoryImpl) FetchAppsByEnvironment(appListingFilter hel
 	return appEnvArr, nil
 }
 
-// It will return the deployment detail of any cd pipeline which is latest triggered for Environment of any App
-func (impl AppListingRepositoryImpl) DeploymentDetailsByAppIdAndEnvId(ctx context.Context, appId int, envId int) (bean.DeploymentDetailContainer, error) {
+func (impl AppListingRepositoryImpl) FetchAppsByEnvironmentV2(appListingFilter helper.AppListingFilter) ([]*bean.AppEnvironmentContainer, int, error) {
+	impl.Logger.Debug("reached at FetchAppsByEnvironment:")
```
Comment (kripanshdevtron) [inline internal/sql/repository/AppListingRepository.go:240]: we should print filter metadata

File: internal/sql/repository/AppListingRepository.go [hunk 6]
```diff
@@ -200,38 +236,153 @@ func (impl AppListingRepositoryImpl) FetchAppsByEnvironment(appListingFilter hel
...
+	var appEnvArr []*bean.AppEnvironmentContainer
+	appsSize := 0
+	lastDeployedTimeMap := make(map[int]string)
+	var appEnvContainer []*bean.AppEnvironmentContainer
+	var lastDeployedTimeDTO = make([]*bean.AppEnvironmentContainer, 0)
+
+	if string(appListingFilter.SortBy) == helper.LastDeployedSortBy {
+
+		query := impl.appListingRepositoryQueryBuilder.GetAppIdsQueryWithPaginationForLastDeployedSearch(appListingFilter)
+		impl.Logger.Debugw("basic app detail query: ", query)
```
Comment (kripanshdevtron) [inline internal/sql/repository/AppListingRepository.go:250]: need to correct log msg

File: pkg/cluster/ClusterService.go [hunk 7]
```diff
@@ -455,7 +497,52 @@ func (impl *ClusterServiceImpl) Update(ctx context.Context, bean *ClusterBean, u
 	if bean.HasConfigOrUrlChanged && util2.IsBaseStack() {
 		impl.SyncNsInformer(bean)
 	}
-	return bean, err
+	impl.logger.Infow("saving secret for cluster informer")
```
Comment (kripanshdevtron) [inline pkg/cluster/ClusterService.go:500]: add some metadata here in log

File: pkg/cluster/ClusterService.go [hunk 8]
```diff
@@ -455,7 +497,52 @@ func (impl *ClusterServiceImpl) Update(ctx context.Context, bean *ClusterBean, u
...
+	restConfig := &rest.Config{}
+
+	restConfig, err = rest.InClusterConfig()
+	if err != nil {
+		impl.logger.Errorw("Error in creating config for default cluster", "err", err)
+		return bean, nil
+	}
+
+	httpClientFor, err := rest.HTTPClientFor(restConfig)
```
Comment (kripanshdevtron) [inline pkg/cluster/ClusterService.go:509]: consider OTEL included common method here

File: pkg/cluster/ClusterService.go [hunk 9]
```diff
@@ -220,6 +232,36 @@ func (impl *ClusterServiceImpl) Save(parent context.Context, bean *ClusterBean,
...
+	//creating cluster secret, this secret will be read informer in kubelink to know that a new cluster has been added
+	secretName := fmt.Sprintf("%s-%v", SECRET_NAME, bean.Id)
+
+	data := make(map[string][]byte)
+	data[SECRET_FIELD_CLUSTER_ID] = []byte(fmt.Sprintf("%v", bean.Id))
+	data[SECRET_FIELD_ACTION] = []byte(CLUSTER_ACTION_ADD)
+	data[SECRET_FIELD_UPDATED_ON] = []byte(time.Now().String()) // this field will ensure that informer detects change as other fields can be constant even if cluster config changes
+	_, err = impl.K8sUtil.CreateSecret(DEFAULT_NAMESPACE, data, secretName, CLUSTER_MODIFY_EVENT_SECRET_TYPE, k8sClient)
+	if err != nil {
+		impl.logger.Errorw("error in updating secret for informers")
```
Comment (kripanshdevtron) [inline pkg/cluster/ClusterService.go:261]: error msg is not logged here.

File: pkg/cluster/ClusterServiceExtended.go [hunk 10]
```diff
@@ -392,5 +394,25 @@ func (impl ClusterServiceImplExtended) DeleteFromDb(bean *ClusterBean, userId in
...
+		return nil
+	}
+	k8sClient, err := v12.NewForConfigAndClient(restConfig, httpClientFor)
+	if err != nil {
+		impl.logger.Errorw("error creating k8s client", "error", err)
+		return nil
+	}
+	secretName := fmt.Sprintf("%s-%v", "cluster-event", bean.Id)
+	err = impl.K8sUtil.DeleteSecret("default", secretName, k8sClient)
+	impl.logger.Errorw("error in deleting secret", "error", err)
```
Comment (kripanshdevtron) [inline pkg/cluster/ClusterServiceExtended.go:416]: err condition is missing here

File: pkg/pipeline/CdHandler.go [hunk 11]
```diff
@@ -1135,50 +1136,56 @@ func (impl *CdHandlerImpl) FetchAppWorkflowStatusForTriggerViewForEnvironment(en
...
-
 		pipelineIds = append(pipelineIds, pipeline.Id)
+		pipelineAppMap[pipeline.Id] = pipeline.AppId
 	}
 	//authorization block ends here
+	if len(pipelineIds) == 0 {
+		return nil, nil
+	}
 
 	if len(pipelineIds) == 0 {
```
Comment (kripanshdevtron) [inline pkg/pipeline/CdHandler.go:1183]: do we need this check if above block is present

**Review by vikramdevtron — DISMISSED** [review]

File: api/module/ModuleRouter.go (hunk 1 above)
Comment (prakash100198) [reply api/module/ModuleRouter.go:20]: Tested the api route, golang didn't take queries as optional, but I wanted to make it optional, that's why I had put it there

**Review by vikramdevtron — APPROVED** [review]

**Review by vikramdevtron — APPROVED** [review]

**Review by prakash100198 — APPROVED** [review]

**Pull request merged**